### DISCUSSION TITLE
PRECOMPILE::ecrecover

### DISCUSSION THREAD

**Discussion by omeguhh** [foundry-rs/foundry#2214] [Help] [answered]
I'm testing this function:

```solidity
function testCanBeRefundedBeforeEndTime() public {
        grantFunder.grantRole(0x4f506ac322e467a10b006ee5ecfc5b7781c5c43d2ffbc468de868900d27b9945, funder);
        vm.deal(funder, 3 ether);
        vm.deal(address(grantFunder), 10 ether);
        emit log_uint(funder.balance);
        grantFunder.createGrant(2, recipient);
        vm.prank(funder);
        grantFunder.deposit{value: 1 ether}(1 ether, 1);
        vm.prank(funder);
        //emit log_uint(address(grantFunder).balance);
        grantFunder.reclaimDeposit(1);
        emit log_uint(funder.balance);
        console.log(address(grantFunder));
    }
```
Which calls this function:

```solidity
function reclaimDeposit(uint256 grantNo) payable external onlyRole(FUNDER) {
        require(msg.sender != address(0), "This is not a valid address.");
        Grant storage grant = grantById[grantNo];
        // Remember to add in || in case block.timestamp > endTime but goal hasn't been reached. 
        require((grant.endTime >= block.timestamp && grant.goal > grant.grantBalance) || grant.goal > grant.grantBalance, "This campaign has ended and has met its fundraising goal.");
        require(contributions[grantNo][msg.sender] > 0, "You have not contributed to this campaign.");
        uint256 _amount = contributions[grantNo][msg.sender];
        contributions[grantNo][msg.sender] = 0;
        grant.grantBalance -= _amount;
        payable(msg.sender).transfer(_amount);
    }
```

And I am getting this error:

<img width="456" alt="Screen Shot 2022-07-05 at 8 16 51 AM" src="https://user-images.githubusercontent.com/105019014/177336712-99fcb33d-f591-4328-b25f-dcd0fa4a2bff.png">

However the user calling the function has the required roles (FUNDER). My contract works when tested in remix so Im not sure what the issue is here. Additionally, I tested `funder` address and switched it from `address(1)` to `address(grantFunder)` and when it was set to the address of the contract it passed. 

Any help with this would be great. Thanks!

UPDATE: I continued testing along and created a similar test to test Claiming a Grant as a grant recipient and received this error:
<img width="641" alt="Screen Shot 2022-07-05 at 8 42 52 AM" src="https://user-images.githubusercontent.com/105019014/177341996-7ef54a73-77cf-417d-a126-78a0bd7cf456.png">

Happy to provide any additional context as needed. Thanks!

**Comment by onbjerg** [comment, marked as answer]
What addresses are you using? Generally, you should only use addresses above `address(9)`, since the others are precompiles. See also https://github.com/foundry-rs/foundry/issues/1228 and https://github.com/foundry-rs/foundry/issues/1159

**Reply by omeguhh** [reply]
Hey there again ser thanks for your help on that - I definitely didn't know and that helped. Super helpful ive been stuck on this all morning haha

**Reply by onbjerg** [reply]
Yeah it's not immediately obvious which is why we are going to disallow those addresses for some cheatcodes. The only issue is that precompiles on other networks might occupy different addresses, so we'll have to think about that a bit :smile:
